### PR TITLE
chore: support simpleauth cd on GitHub action

### PR DESCRIPTION
**Pull request by wenytang-ms**

### CODE REVIEW

File: .github/scripts/simpleauth-azdo-pipelines.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+x=1
+restUrl="https://dev.azure.com/mseng/VSIoT/_apis/build/latest/$2?api-version=6.0-preview.1"
```
Comment (blackchoey) [inline .github/scripts/simpleauth-azdo-pipelines.sh:3]: is there non-preview api available?

File: .github/scripts/simpleauth-azdo-pipelines.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+x=1
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply .github/scripts/simpleauth-azdo-pipelines.sh:2]: updated

File: .github/scripts/simpleauth-azdo-pipelines.sh (hunk 1 above)
Comment (wenytang-ms) [reply .github/scripts/simpleauth-azdo-pipelines.sh:3]: copy this api from https://docs.microsoft.com/en-us/rest/api/azure/devops/pipelines/runs/run-pipeline?view=azure-devops-rest-6.0

File: .github/scripts/update-simpleauth-ver.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        throw err;
+    }
+
+    // convert XML data to JSON object
+    xml2js.parseString(data, (err, result) => {
+        if (err) {
+            throw err;
+        }
+
+        // replace `Neo4j` with `ArangoDB`
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply .github/scripts/update-simpleauth-ver.js]: sure

File: .github/scripts/update-simpleauth-ver.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        result['Project'].PropertyGroup[1].Version[0] = simpleauthVer;
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply .github/scripts/update-simpleauth-ver.js]: sure, upadte

**Review by LongOddCode — APPROVED** [review]

**Review by blackchoey — APPROVED** [review]

**Comment by blackchoey** [comment]
I saw Simple Auth CI failed, please take a look before merge

File: .github/scripts/simpleauth-azdo-pipelines.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+countNum=1
+restUrl="https://dev.azure.com/mseng/VSIoT/_apis/build/latest/$2?api-version=6.0-preview.1"
+rsp=$(curl -u :$1 $restUrl | jq -r '.value| .[0]')
+status=$(echo $rsp | jq -r '.state')
+buildId=$(echo $rsp | jq -r '.id')
```
Comment (wenytang-ms) [inline .github/scripts/simpleauth-azdo-pipelines.sh:6]: this step is get latest build pipeline id and status, if status is completed, we can get asset url leverage the build id

File: .github/scripts/simpleauth-azdo-pipelines.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+echo "===== build id is: " $buildId
+echo "===== build pipeline status: " $status
+while [[ $countNum -le 50 && "$status" != "completed" ]]
+do 
+    sleep 1m
+    rsp=$(curl -u :$1 $restUrl | jq -r '.value| .[0]')
+    status=$(echo $rsp | jq -r '.state')
+    echo "loop status" $status
+    countNum=$(( $countNum + 1 ))
```
Comment (wenytang-ms) [inline .github/scripts/simpleauth-azdo-pipelines.sh:15]: in loop, every time sleep 1m, and fetch pipeline status, if status is completed or exceed 50 times, break

File: .github/scripts/simpleauth-azdo-pipelines.sh [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+done
+if [[ "$status" != "completed" ]]
+then
+exit 1
+fi
```
Comment (wenytang-ms) [inline .github/scripts/simpleauth-azdo-pipelines.sh:20]: if timeout return exit1

File: .github/scripts/simpleauth-azdo-pipelines.sh [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+restUrl="https://dev.azure.com/mseng/VSIoT/_apis/build/builds/$buildId/artifacts?api-version=6.0"
+asset_rsp=$(curl -u :$1 $restUrl)
+echo "====== asset url response:" $asset_rsp
+asset_id=$(echo $asset_rsp | jq '.value |.[] | .resource' |jq '.data' | tr -d -c 0-9)
+echo "====== asset id is: " $asset_id
```
Comment (wenytang-ms) [inline .github/scripts/simpleauth-azdo-pipelines.sh:26]: get the asset_id to concat an url to download the target file

File: .github/scripts/update-simpleauth-ver.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
+const fs = require("fs");
+const path = require("path")
+const simpleauth = path.join(__dirname, "../../packages/simpleauth")
+const xml2js = require(path.join(simpleauth, "node_modules/xml2js"))
+const csprojFile = path.join(simpleauth, "src/TeamsFxSimpleAuth/Microsoft.TeamsFx.SimpleAuth.csproj");
+const simpleauthVer = require(path.join(simpleauth, "package.json")).version
```
Comment (wenytang-ms) [inline .github/scripts/update-simpleauth-ver.js:6]: csproj file is an xml file.

File: .github/workflows/cd.yml [hunk 10]
```diff
@@ -55,20 +55,15 @@ jobs:
         with:
           node-version: 14
 
+      - name: Setup .NET
+        uses: actions/setup-dotnet@v1
+        with:
+          dotnet-version: 3.1.x
+
```
Comment (wenytang-ms) [inline .github/workflows/cd.yml:62]: cause alpha and rc CD needs to run dotnet build, so install here

File: .github/workflows/cd.yml [hunk 11]
```diff
@@ -179,6 +175,56 @@ jobs:
       - name: cleanup templates
         run: rm -rf ./*.zip && git clean -df
 
+      - name: trigger simpleauth azure build pipeline for stable release
```
Comment (wenytang-ms) [inline .github/workflows/cd.yml:178]: only run stable CD, pipeline will trigger the AZDO pipeline

File: .github/workflows/cd.yml [hunk 12]
```diff
@@ -179,6 +175,56 @@ jobs:
...
+          azure-pipeline-name: 'Build Simple Auth Release Artifacts' 
+          azure-devops-token: '${{ secrets.ADO_PAT }}'
+      
+      - name: check build status and download assets for stable release
+        if: ${{ contains(steps.version-change.outputs.CHANGED, 'simpleauth') && github.event_name == 'workflow_dispatch' && github.event.inputs.preid == 'stable' }}
+        run: |
+          output=$(./.github/scripts/simpleauth-azdo-pipelines.sh ${{ secrets.ADO_PAT }} 11303 >&1)
+          if [ -z $output ]; then exit 1; fi
+          echo $output
+          asset_id=$(echo $output | rev |cut -d ' ' -f 1 | rev)
```
Comment (wenytang-ms) [inline .github/workflows/cd.yml:192]: get the latest info about the assets_id number

File: .github/workflows/cd.yml [hunk 13]
```diff
@@ -179,6 +175,56 @@ jobs:
...
+          simpleAuthVer=$(echo ${{ steps.version-change.outputs.CHANGED }} | grep simpleauth | cut -d '@' -f2)
+          curl "https://dev.azure.com/mseng/_apis/resources/Containers/$asset_id/drop?itemPath=drop/Microsoft.TeamsFx.SimpleAuth_$simpleAuthVer.zip" -L -O -u :${{ secrets.ADO_PAT }}
```
Comment (wenytang-ms) [inline .github/workflows/cd.yml:194]: download simpleauth bits.

File: .github/workflows/cd.yml [hunk 14]
```diff
@@ -179,6 +175,56 @@ jobs:
...
+          if [ ! -f Microsoft.TeamsFx.*.zip ]; then exit 1; fi
+      
+      - name: build SimpleAuth on prerelease
+        if: ${{ contains(steps.version-change.outputs.CHANGED, 'simpleauth') && github.event.inputs.preid != 'stable' }}
```
Comment (wenytang-ms) [inline .github/workflows/cd.yml:198]: CD for alpha and rc, simpleauth can build local, and there no need to trigger the AZDO pipeline

File: .github/workflows/cd.yml [hunk 15]
```diff
@@ -179,6 +175,56 @@ jobs:
...
+        run: |
+          dotnet build --configuration Release
+          simpleAuthVer=$(echo ${{ steps.version-change.outputs.CHANGED }} | grep simpleauth | cut -d '@' -f2)
+          cd src/TeamsFxSimpleAuth/bin/Release/*
+          zip -r Microsoft.TeamsFx.SimpleAuth_$simpleAuthVer.zip .
+          mv Microsoft.TeamsFx.SimpleAuth_*.zip ../../../../../../..
```
Comment (wenytang-ms) [inline .github/workflows/cd.yml:204]: mv simpelauth bits to root dir.

File: .github/workflows/cd.yml [hunk 16]
```diff
@@ -179,6 +175,56 @@ jobs:
...
+          retry_on: error
+          command: |
+            if [ -f Microsoft.TeamsFx.SimpleAuth_*.zip ]; then
+              cp Microsoft.TeamsFx.SimpleAuth_*.zip packages/fx-core/templates/plugins/resource/simpleauth/SimpleAuth.zip
+            else
+              ./.github/scripts/downloadSimpleAuth.ps1
+            fi
+
+      - name: publish simpleauth to github release page
+        if: ${{ contains(steps.version-change.outputs.CHANGED, 'simpleauth') && github.ref == 'refs/heads/main' }}
```
Comment (wenytang-ms) [inline .github/workflows/cd.yml:221]: publish simpleauth zip bits to github release page

**Pull request merged**